### ISSUE TITLE
[Bug]: mpr_refine_portal has no iteration cap; a non-converging portal spins a GPU lane forever (standalone reproduction attached)

### ISSUE THREAD

**Issue by chkxw** [Genesis-Embodied-AI/genesis-world#3314]
## Bug Description

`mpr_refine_portal` (`genesis/engine/solvers/rigid/collider/mpr.py:263-310`, the `while True` at line 280) has no iteration cap. Its sibling `mpr_find_penetration` is capped by `CCD_ITERATIONS`, and libccd caps both loops. On near-degenerate geometry the portal keeps expanding by rounding-sized steps without encapsulating the origin or reaching the tolerance: one lane spins forever, the kernel never returns, and the host blocks at its next device synchronisation. No NaN or blown-up state is involved, and the point at which it happens is deterministic given the trajectory.

Observed on a dual-quadruped push (two Go1 URDFs, a box, a plane), 256–512 envs, both constraint-solver arms, on RTX 5090, RTX 5070, L40S, A100-PCIE-40GB and A100-SXM4-80GB, with genesis-world 1.3.3 and `main` (01387ec8). `cuda-gdb` on the frozen process shows one resident kernel (`_func_narrowphase_multicontact…` or `_func_narrowphase_contact0`), one active lane inside `mpr_refine_portal` reached through `func_mpr_contact`, its 31 warp-mates parked. With the loop bounded, the same trajectory records a cap exactly in the window where the unbounded build stops: A100, seed 1 — stock stops at PPO update 49, the cap fires at per-env step 6384 inside the (6272, 6400] window, the run completes 190 updates with 7 caps; seed 2 — stock stops at update 5, first cap at step 656, 20 caps in 190 updates; RTX 5090 — same picture on both arms.

A second uncapped loop of the same class: the three `while is_improving*` chases in `multi_contact.py::func_approximate_polygon_with_quad` (contact-patch path only). One NaN vertex makes every test false and the lane cycles forever.

## Steps to Reproduce

Standalone, no trainer, built on #3309's `save_checkpoint` / `load_checkpoint`. Files: https://github.com/chkxw/genesis-world/releases/tag/repro-3314 (~177 MB: `ckpt_512.gsckpt`, `ops.pkl.gz`, `meta.json`, `state_before_step512.npz`, `repro_3314_ckpt_replay.py`, `0007-pr3309-read-frame-zerocopy-fallback.patch`, `REPRO_3314.md`). A 256-env scene checkpointed right before physics step 512, plus the recorded entity calls for the following steps.

Branch as is:

```
git clone https://github.com/Genesis-Embodied-AI/genesis-world && cd genesis-world && git checkout 01387ec8
git apply 0007-pr3309-read-frame-zerocopy-fallback.patch   # #3309 checkpoint I/O with zero-copy on CUDA; reported on #3309
pip install -e .
python repro_3314_ckpt_replay.py <files dir> --ckpt-step 512 --steps 256 --timeout 120
```

Output: the restored state matches the snapshot (`max|diff|=0` for all six entities), state lines for steps 512, 528 and 544, then `HANG` — the GPU at 100 % with the lane in `mpr_refine_portal`.

With the bound:

```
git fetch https://github.com/chkxw/genesis-world fix/mpr-portal-discovery-cap && git cherry-pick a0f78f7e f018e622
python repro_3314_ckpt_replay.py <files dir> --ckpt-step 512 --steps 256 --timeout 120
```

Output: all 256 steps in about 3 s, `capped_envs=[179]` from step 560, `COMPLETED`; the printed state equals the recording's up to that portal.

Recorded on a single RTX 5090 (driver 595.84). Whether another architecture walks into the same portal from this checkpoint is untested: the collider is bit-deterministic on one card, not across cards. Resuming an exact `TrajectoryFile` (#3309) from frame 512 gives the same two outcomes.

## Expected Behavior

Both loops terminate for every input. A non-converging portal, or a NaN polygon, is dropped and reported per env, never spun on.

## Suggested fix

Fork branch `fix/mpr-portal-discovery-cap` (two commits; was #3315): cap portal discovery with `CCD_ITERATIONS` like refinement and return "no portal"; set a non-raising per-env `ErrorCode.UNCONVERGED_CONTACT_PORTAL`, exposed through `get_error_envs_mask()`, right after the attempt that hit the cap (a warm-start retry otherwise erases it); cap each quad chase at `nverts` and drop that pair's manifold. A NaN-aware comparison is not an option for the quad chase because the kernels are compiled with `NoNaNsFPMath`.

## Environment

genesis-world 1.3.3 (PyPI) and `main` 01387ec8 (+ #3309); quadrants 1.3.0; torch 2.8.0+cu128; Python 3.12; CUDA 12.8; Ubuntu 22.04 / 24.04. Drivers: 595.84 (RTX 5090), 580.173.02 (RTX 5070, L40S, A100-PCIE), 595.58.03 / 595.71.05 (A100-SXM4).

**Comment by chkxw** [comment]
Update on the primary loop, with a negative result stated plainly.

**The pair is known and ordinary.** Recording the inputs of the cap event in the patched build (same trajectory on which stock freezes 3/3 after the first PPO update, RTX 5090, 512 envs, monolith arm) identifies the pair: the Go1 base collision box (0.3762 × 0.0935 × 0.114 m; the robot has fallen and is flipped about 180°) against the pushed box (both BOX primitives), reached through the multicontact kernel's perturbed re-detection. The event is bit-for-bit reproducible across runs on that trajectory. Full float32 inputs (poses, centres, geom ids) are available on request.

**An input-level replay does not reproduce the spin.** Feeding those exact inputs to `func_mpr_contact_from_centers` from a one-thread kernel converges on CPU (stock and bounded) and on the GPU (stock): contact found, penetration 0.056, portal EXACT, no cap. So whether the discovery loop terminates depends on the full narrowphase kernel's own arithmetic (fast-math contraction and scheduling differ with code shape) and possibly on the pair's warm-start direction from the contact cache, not on the inputs alone. An exact standalone reproducer would have to drive the real narrowphase kernel on a scene with the pair at its unperturbed poses and a pre-seeded cache; we stopped there.

What stands as the reproduction is the training-run before/after on identical trajectories (stock freezes; the bounded loop records one dropped portal in one env and continues), on RTX 5090 with both solver arms, and the deterministic freeze points on A100 (same seed, same PPO update, three allocations). The bound in #3315 fixes it regardless of which arithmetic path a given chip takes.

Also for the record: a device fence at the PPO-update boundary plus an audit of the trainer/simulator handoff (zero mismatches over 16 updates) did not prevent the freeze on the RTX 5090 either; every freeze and every recorded cap event sits at the first PPO boundary after a periodic environment reset, i.e. the near-degenerate contact comes from reset/spawn poses.

**Comment by chkxw** [comment]
Confirmation on A100 (Ampere), with the bound from #3315 applied to genesis-world 1.3.3.

Hardware: NVIDIA A100-PCIE-40GB, driver 580.173.02, sm_80. Workload: two Go1 quadrupeds pushing a box, 512 envs, PPO, `use_deterministic_algorithms=True` (decomposed constraint arm). Same seed and configuration as four stock runs that hang.

**Stock 1.3.3 hangs at a fixed point of the trajectory.** Four independent allocations on three nodes stopped after exactly 49 PPO updates, with `nvidia-smi pmon` showing the process at 99 % SM (spinning in-kernel, not starved) and the solver reporting no error envs and no NaN. Two of those runs stopped in a bit-identical state (per-env step 6272, max|q| = 5.312 / max|v| = 7.093 and 6.303 / 6.848 for the two robots, 12 contacts) — those two were on the same node, so that part is reproducibility across allocations on one node; the update-49 result spans three nodes.

**With the bound applied, the same run reaches the same state and continues:**

```
step=6384  err_envs=1 which=[182]  portal_cap_hits=1   <- cap fires, no exception
step=6512  err_envs=0 which=[]     portal_cap_hits=1   <- update 50, still running
```

The capped contact sets the non-raising per-env bit for one env (182); that env's next reset clears the bit while the cumulative counter stays at 1. No scene reset, no exception, no NaN.

Note on step numbers: the stock runs sampled state every 128 steps, so their last sample was 6272 and the hang lies in (6272, 6400]; the bounded run sampled every 16 steps and the cap fired at 6384, inside that window. The bound fires where the stock build dies. This was a pre-registered prediction (cap event at update 49, run continues); all three of its falsifiers stayed clear. One patched arm so far; two more are queued and will be reported here.

**Comment by chkxw** [comment]
Before/after on Ampere with the fix from #3315 (as a patch on genesis-world 1.3.3, quadrants 1.3.0, plus a
per-slot cap counter for verification), same workload as the report, same seed, A100-PCIE-40GB, driver
580.173.02, 512 envs, deterministic mode (decomposed constraint-solver arm, host do-while loop).

**Stock 1.3.3, seed 1.** Four allocations on three nodes all stopped at PPO update 49: the last 128-step
sample is per-env step 6272 (`max|q| = 5.312`, `max|v| = 7.093`, no error flag, no NaN, identical across the
runs) and the freeze lies inside `(6272, 6400]`; `nvidia-smi pmon` shows the process at 99 % SM. A different
contact model on the same seed (Signorini, `sig_s1`) froze at update 55, step 7040, likewise at 99 % SM
with clean state: the model moves the point, it does not remove it.

**Patched, seed 1 (`pdef_s1`, job 64018218, 16-step sampling).** Trajectory bit-identical to stock through
step 6272, then

```
G13STATE step=6384 err_envs=1 which=[182] cap_envs=1 other_err_envs=0 mpr_cap_hits=1   <- portal cap
G13STATE step=6512 err_envs=0 which=[]    cap_envs=1 other_err_envs=0 mpr_cap_hits=1   <- update 50, continuing
```

i.e. the first cap fires inside the window where stock dies, in one env (182), the env's flag is cleared at
its next reset, no exception, and the run goes on. It completed all 190 updates (24 320 per-env steps, with
the completion marker). Seven portal caps in total, each in a different env and each followed by normal
stepping:

| per-env step | env |
|---|---|
| 6384 | 182 |
| 6656 | 216 |
| 7136 | 103 |
| 13648 | 298 |
| 16048 | 349 |
| 18048 | 133 |
| 21472 | 404 |

Roughly one cap per 27 updates, one per ~3500 steps at 512 envs, on this workload. Two independent counters
agree: the number of per-env `errno` 0 -> 1 transitions seen at 16-step sampling (a lower bound, since a
reset clears the bit) and a never-cleared per-slot counter incremented inside `mpr_refine_portal` both read
exactly 7. Environment verified as the bounded build (`ErrorCode` bit present, counter present, cap in
`mpr.py`); a manifest label in that job said an older patch revision and was corrected.

This is the same picture as the RTX 5090 before/after already in the issue (monolith arm, 512 envs: stock
froze after update 1 in 3 of 3 runs, patched recorded one cap in one env in that interval and completed;
decomposed arm with the device conditional graph: one cap at step 4096 after 31 clean updates, run
continued). Ampere, host-loop path, deterministic mode: same loop, same fix, no exception, no whole-scene
reset.

**Comment by chkxw** [comment]
Seed 2 of the Ampere pair (same hardware and configuration as above: A100-PCIE-40GB, driver 580.173.02, 512 envs, deterministic mode).

**Stock 1.3.3:** stops after PPO update 5 (`def_s2`; 99 % SM, no error envs, no NaN).

**Patched (`pdef_s2`):** completes all 190 updates. 20 portal caps in 20 distinct environments, at per-env steps 656, 688, 1600, 2176, 5152, 9600, 9760, 11584 (×2), 11824, 13056, 16832, 17072, 17184, 17312, 18240, 21024, 22864, 23104, 24128. The first cap is at step 656 = update 5.1, i.e. where the stock run stops; on seed 1 the first cap was at 6384, inside the window where that seed's stock run stops. On both seeds the bounded loop fires at the point where the unbounded build hangs, and the run continues.

Cap frequency follows the trajectory, not the machine: about one per 27 updates on seed 1, one per 9.5 on seed 2 — and seed 2 is the seed whose stock run stops at update 5 rather than 49. The per-slot counter reads 20; the sampled per-env `errno` transition count reads 19 because two caps fell inside one 16-step sampling interval (the counter jumps from 7 to 9 between consecutive samples), which is the documented lower-bound behaviour of that check.

**Comment by chkxw** [comment]
Here is a reproduction that needs no trainer. It is built on #3309's `Scene.save_checkpoint` / `load_checkpoint`
(exact mode): a 256-env scene (two Go1 URDFs, a box, a plane; `use_deterministic_algorithms=True`, dt 0.04,
4 substeps) checkpointed right before physics step 512, plus the entity calls (PD targets, resets) the training loop
made from there on, replayed step by step by a small script. `main` + #3309 as is freezes within 48 steps of the
checkpoint; with the portal-discovery cap the same window completes and exactly one env is flagged.

**Hardware / software.** One machine only: a single RTX 5090 (sm_120), driver 595.84 (open kernel module), Linux,
Python 3.12, torch 2.8.0+cu128, quadrants 1.3.0, genesis-world `main` + #3309 at 01387ec8. The hang was first seen on
RTX 5070 and L40S with 1.3.3 (#3310 and above); this checkpoint is the 5090 trajectory. Whether the same file walks
into the same portal on another architecture is untested (the collider is bit-deterministic on one card, not across
cards).

**Files** (release https://github.com/chkxw/genesis-world/releases/tag/repro-3314):

| file | bytes | content |
|---|---|---|
| `ckpt_512.gsckpt` | 159,542,570 | `Scene.save_checkpoint` right before step 512 |
| `ops.pkl.gz` | 17,395,789 | the recorded entity calls, by step (`control_dofs_position` targets; resets before the window) |
| `meta.json` | 3,361 | `gs.init` kwargs and environment of the recording |
| `state_before_step512.npz` | 135,370 | qpos / dof velocity snapshot right before step 512, to check the restored state |
| `repro_3314_ckpt_replay.py` | 5,991 | the replay script (numpy, torch, genesis only) |
| `0007-pr3309-read-frame-zerocopy-fallback.patch` | 3,562 | needed on the #3309 branch to write/read checkpoints with `gs.use_zerocopy=True` on CUDA (separate comment on #3309) |

**Branch as is (expect HANG):**

    git clone https://github.com/Genesis-Embodied-AI/genesis-world && cd genesis-world && git checkout 01387ec8
    git apply 0007-pr3309-read-frame-zerocopy-fallback.patch
    pip install -e .
    python repro_3314_ckpt_replay.py <files dir> --ckpt-step 512 --steps 256 --timeout 120

Output: `fidelity restored-state e{1..6}_qpos/vel: max|diff|=0` for all six entities, the state lines for steps 512,
528 and 544, then nothing until `HANG: the replay did not finish within 120s (child killed)`; the GPU sits at 100 %
with the frozen lane in `mpr_refine_portal` (the `while True` at `collider/mpr.py:280`).

**With the cap (expect COMPLETED, env 179 flagged):** the fork branch `chkxw/genesis-world:fix/mpr-portal-discovery-cap`
carries two commits on top of `main`: a0f78f7e bounds the discovery loop by `CCD_ITERATIONS` like the refinement loop
and reports the env through `get_error_envs_mask()`; f018e622 raises that flag right after the attempt that hit the
cap (a warm-start retry otherwise erases it). Cherry-pick or take whatever shape you prefer:

    git fetch https://github.com/chkxw/genesis-world fix/mpr-portal-discovery-cap
    git cherry-pick a0f78f7e f018e622
    python repro_3314_ckpt_replay.py <files dir> --ckpt-step 512 --steps 256 --timeout 120

Output: the same restore, all 256 steps in about 3 s, `err_envs=1 capped_envs=[179]` from step 560 on, and
`COMPLETED: 256 steps replayed from the checkpoint; envs whose portal hit the iteration cap: [179]`. The printed
max-q / max-v lines equal the recording's, so the cap changes nothing until the one portal that never converges.

**Trajectory-file route** (local validation only: an exact 256-env frame is ~5.8 MB, the file 3.1 GB): the branch as
is was recorded into an exact `TrajectoryFile` (chunk_size=16, max_size raised) until it hung; 544 frames are readable
(the chunk in flight at the hang is lost, as designed). `load_trajectory` + `seek(512)` restores exactly the state the
checkpoint holds (bit-identical qpos/velocity), the recording's own control log equals the checkpoint recording's for
steps 0..511, and the two verdicts repeat: branch as is hangs after step 544, cap completes with env 179 flagged from
step 560. From the last readable frame (543) the branch as is also hangs, but one window later, in (559, 575]: an exact
frame restores INFO/STATE/WARMSTART/DERIVED and not SCRATCH, while the checkpoint's terminal record includes scratch,
so a resume from a frame is not bit-exact when some scratch array is read before it is written in a step (31 steps from
frame 512 re-warm it, 1 step from frame 543 does not). That is a #3309 fidelity note; the checkpoint is the
reproduction to use.

**Comment by duburcqa** [comment]
@chkxw I will have a look asap!

**Comment by duburcqa** [comment]
Reproduced!

**Comment by chkxw** [comment]
> Reproduced!

Thanks for confirming so quickly!

**Comment by duburcqa** [comment]
@chkxw Let me know if the issue is fixed on your side!

**Comment by chkxw** [comment]
Works well our side, thanks for the quick turnaround. On genesis-world 1.4.0 from PyPI, the released checkpoint replays its 256-step window on an RTX 5090 with no error env, and the two training configurations that used to freeze on 1.3.3 now complete, thanks alot
